### PR TITLE
PWGHF: Add final version Lb tree creator

### PR DESCRIPTION
**Pull request by mjongerh**
Adding final version of the Lb tree creator including TOF and (f)RICH for final state particles

Run3AnalysisValidation: https://github.com/AliceO2Group/Run3Analysisvalidation/pull/351

### CODE REVIEW

File: PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+
+  void process(aod::Collisions const& collisions,
+               aod::McCollisions const& mccollisions,
+               soa::Join<aod::HfCandLb, aod::HfCandLbMCRec, aod::HFSelLbToLcPiCandidate> const& candidates,
+               soa::Join<aod::HfCandProng3, aod::HfCandProng3MCRec, aod::HFSelLcCandidate> const& Lccandidates,
+               soa::Join<aod::McParticles, aod::HfCandLbMCGen> const& particles,
+               aod::BigTracksMC const& bigtracksmc,
+               ExtendedTracksPID const&,
+               aod::FRICHs const&,
+               aod::RICHs const&)
```
Comment (vkucera) [inline PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx:230]: Many unused tables
```suggestion
  void process(soa::Join<aod::HfCandLb, aod::HfCandLbMCRec, aod::HFSelLbToLcPiCandidate> const& candidates,
               soa::Join<aod::HfCandProng3, aod::HfCandProng3MCRec, aod::HFSelLcCandidate> const&,
               TracksExtendedPID const&,
               aod::FRICHs const&,
               aod::RICHs const&)
```

File: PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+  Produces<o2::aod::HfCandLbFull> rowCandidateFull;
+  Produces<o2::aod::HfCandLbFullEvents> rowCandidateFullEvents;
+  Produces<o2::aod::HfCandLbFullParticles> rowCandidateFullParticles;
+
+  void init(InitContext const&)
+  {
+  }
+
+  using TracksPID = soa::Join<aod::BigTracksPID, aod::HfTrackIndexALICE3PID>;
+  using ExtendedTracksPID = soa::Join<TracksPID, aod::TracksExtended>;
```
Comment (vkucera) [inline PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx:220]: You are using only the second one.
```suggestion
  using TracksExtendedPID = soa::Join<aod::BigTracksPID, aod::HfTrackIndexALICE3PID, aod::TracksExtended>;
```

File: PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+          auto fRICHTrk2K = -5000.0;
+          auto fRICHTrk3Pi = -5000.0;
+          auto fRICHTrk3p = -5000.0;
+
+          if (track0.has_rich())
+            RICHPi0 = track0.rich().richNsigmaPi();
+          if (track1.has_rich())
+            RICHTrk1Pi = track1.rich().richNsigmaPi();
+          if (track1.has_rich())
+            RICHTrk1p = track1.rich().richNsigmaPr();
```
Comment (vkucera) [inline PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx:266]: Merge

File: PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+          if (track2.has_rich())
+            RICHTrk2K = track2.rich().richNsigmaKa();
+          if (track3.has_rich())
+            RICHTrk3Pi = track3.rich().richNsigmaPi();
+          if (track3.has_rich())
+            RICHTrk3p = track3.rich().richNsigmaPr();
```
Comment (vkucera) [inline PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx:272]: Merge

File: PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+          if (track1.has_frich())
+            fRICHTrk1Pi = track1.frich().frichNsigmaPi();
+          if (track1.has_frich())
+            fRICHTrk1p = track1.frich().frichNsigmaPr();
+          if (track2.has_frich())
+            fRICHTrk2K = track2.frich().frichNsigmaKa();
+          if (track3.has_frich())
+            fRICHTrk3Pi = track3.frich().frichNsigmaPi();
+          if (track3.has_frich())
+            fRICHTrk3p = track3.frich().frichNsigmaPr();
```
Comment (vkucera) [inline PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx:285]: Same as for RICH

File: PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+          rowCandidateFull(
+            candidate.rSecondaryVertex(),
+            candidate.decayLength(),
+            candidate.decayLengthXY(),
+            candidate.decayLengthNormalised(),
+            candidate.decayLengthXYNormalised(),
+            candidate.chi2PCA(),
+            candidate.impactParameterNormalised0(),
+            candidate.ptProng0(),
+            RecoDecay::P(candidate.pxProng0(), candidate.pyProng0(), candidate.pzProng0()),
```
Comment (vkucera) [inline PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx:296]:
```suggestion
            RecoDecay::p(candidate.pxProng0(), candidate.pyProng0(), candidate.pzProng0()),
```

File: PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+            candidate.impactParameterNormalised1(),
+            candidate.ptProng1(),
+            RecoDecay::P(candidate.pxProng1(), candidate.pyProng1(), candidate.pzProng1()),
```
Comment (vkucera) [inline PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx:299]:
```suggestion
            RecoDecay::p(candidate.pxProng1(), candidate.pyProng1(), candidate.pzProng1()),
```

File: PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+  {
+
+    // Filling candidate properties
+    rowCandidateFull.reserve(candidates.size());
+    for (auto& candidate : candidates) {
+      auto fillTable = [&](int FunctionSelection,
+                           float FunctionInvMass,
+                           float FunctionCt,
+                           float FunctionY) {
+        if (FunctionSelection >= 1) {
```
Comment (vkucera) [inline PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx:240]: Why don't you use a filter?

File: PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+          auto LcCand = candidate.index0_as<soa::Join<aod::HfCandProng3, aod::HfCandProng3MCRec, aod::HFSelLcCandidate>>();
```
Comment (vkucera) [inline PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx:241]:
```suggestion
          auto candLc = candidate.index0_as<soa::Join<aod::HfCandProng3, aod::HfCandProng3MCRec, aod::HFSelLcCandidate>>();
```

File: PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+          auto track1 = LcCand.index0_as<ExtendedTracksPID>();    // granddaughter tracks (lc decay particles)
+          auto track2 = LcCand.index1_as<ExtendedTracksPID>();
+          auto track3 = LcCand.index2_as<ExtendedTracksPID>();
+
+          auto RICHPi0 = -5000.0;
+          auto RICHTrk1Pi = -5000.0;
+          auto RICHTrk1p = -5000.0;
+          auto RICHTrk2K = -5000.0;
+          auto RICHTrk3Pi = -5000.0;
+          auto RICHTrk3p = -5000.0;
```
Comment (vkucera) [inline PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx:252]: Use consistent names
```suggestion
          auto RICHTrk0Pi = -5000.0;
          auto RICHTrk1Pi = -5000.0;
          auto RICHTrk1P = -5000.0;
          auto RICHTrk2K = -5000.0;
          auto RICHTrk3Pi = -5000.0;
          auto RICHTrk3P = -5000.0;
```

File: PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+#include "Framework/AnalysisTask.h"
+#include "DetectorsVertexing/DCAFitterN.h"
+#include "PWGHF/DataModel/HFSecondaryVertex.h"
+#include "PWGHF/DataModel/HFCandidateSelectionTables.h"
+#include "Common/Core/trackUtilities.h"
+#include "ReconstructionDataFormats/DCA.h"
+#include "Common/Core/TrackSelectorPID.h"
+#include "ALICE3/DataModel/RICH.h"
+#include "Common/Core/PID/PIDResponse.h"
+#include "ReconstructionDataFormats/PID.h"
```
Comment (vkucera) [inline PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx:31]: Sort

File: PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+
+/// Writes the full information in an output TTree
+struct HfTreeCreatorLbToLcPi {
+  Produces<o2::aod::HfCandLbFull> rowCandidateFull;
+  Produces<o2::aod::HfCandLbFullEvents> rowCandidateFullEvents;
+  Produces<o2::aod::HfCandLbFullParticles> rowCandidateFullParticles;
+
+  void init(InitContext const&)
+  {
+  }
```
Comment (vkucera) [inline PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx:217]: Is this needed?

**Review by vkucera — CHANGES_REQUESTED** [review]

File: PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx [hunk 13]
```diff
@@ -0,0 +1,361 @@
...
+struct Alice3PidIndexBuilder {
+  Builds<o2::aod::HfTrackIndexALICE3PID> index;
+  void init(o2::framework::InitContext&) {}
+};
+
+/// Writes the full information in an output TTree
+struct HfTreeCreatorLbToLcPi {
+  Produces<o2::aod::HfCandLbFull> rowCandidateFull;
+  Produces<o2::aod::HfCandLbFullEvents> rowCandidateFullEvents;
+  Produces<o2::aod::HfCandLbFullParticles> rowCandidateFullParticles;
```
Comment (vkucera) [inline PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx:213]: You are not filling these tables.

**Review by vkucera — CHANGES_REQUESTED** [review]

File: PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx (hunk 13 above)
Comment (vkucera) [reply PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx]: The table are still there.

Comment (mjongerh) [reply PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx]: That is odd, when I attempt to change the file, these lines are already gone. In the branch the commit was pushed already.
eg: https://github.com/mjongerh/O2Physics/blob/Lb_PRbranch/PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx#L211
https://github.com/AliceO2Group/O2Physics/pull/994/commits/11fbeb3e8155feb833d3441b7e2c22ea5a80c6fd

Comment (vkucera) [reply PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx]: What I meant is that you just removed the two lines but not the definitions of columns and tables.

File: PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx (hunk 10 above)
Comment (vkucera) [reply PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx]: You did not modify it everywhere. The current version will not compile.

File: PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx (hunk 12 above)
Comment (vkucera) [reply PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx:195]: Did you try?

File: PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx (hunk 10 above)
Comment (mjongerh) [reply PWGHF/TableProducer/HFTreeCreatorLbToLcPi.cxx]: Fixed all labels, and checked that it will build and run

**Review by vkucera — APPROVED** [review]

**Pull request merged**